### ISSUE TITLE
feat: include group as argument in @Transform

### ISSUE THREAD

**Issue by lazarljubenovic** [typestack/class-transformer#208]
As excluding and exposing fields can be determined by the group, the way a custom transformation is performed should certainly be dependent on the group as well.

# Current state

Currently the only workaround I can think of is this:

```ts
  @Expose({ groups: ['friend'] })
  get isFriend () {
    return true
  }

  @Expose({ name: 'isFriend', groups: ['stranger', 'me'] })
  get isFriend2 () {
    return false
  }
``` 

# Idea 1

Something like this would be nice:

```ts
@Expose({ groups: ['friend', 'stranger', 'me'] })
@Transform(({ groups }) => groups.some(x => x == 'friend'))
get isFriend () {
  return undefined as boolean
}
```

However, this leaves the awkward useless function body.

# Idea 2

So an even better alternative which relies on reflection but requires some more work:

```ts
@Expose({ groups: ['friend', 'stranger', 'me'] })
isFriend (@Groups() groups: string[]) {
  return groups.some(x => x == 'friend')
}
```

This is the prettiest solution and also gives most flexibility since it's also possible to access other fields as usual in functions.

**Comment by lazarljubenovic** [comment]
I tried to implement the **Idea 2** thingy but I'm having troubles accessing the class itself from here:

https://github.com/typestack/class-transformer/blob/c3cd7f935adaee70c12d89db7327e51ce9c6679a/src/TransformOperationExecutor.ts#L150

The whole function is a bit too huge for me to understand it without some guidance, at least not quick enough. Although I guess we don't really _need_ the decorator there, except it's more developer-friendly API for figuring out where the heck the groups come from. But apart from that, it would be super-easy to just do...

# Idea 3

```ts
@Expose({ groups: ['friend', 'stranger', 'me'] })
isFriend ({ groups }: SomeInterfaceWhichTellsWhatIsGivenToDeveloperToUse) {
  return groups.some(x => x == 'friend')
}
```

In other words, we could just shove in some options to every function and document it. It might not even be that bad.

```ts
// ...
subValue = value[valueKey]({ groups: this.options.groups }); 
// ...
```

# Idea 3b

Heck we could even shove in the whole thing... That might just be the best way after all the decorator talk :sweat_smile: 

```ts
// usage
@Expose({ groups: ['friend', 'stranger', 'me'] })
isFriend ({ groups }: ClassTransformOptions) {
  return groups.some(x => x == 'friend')
}

// implementaion
// ...
subValue = value[valueKey](this.options); 
// ...
```

What do you think?

**Comment by NoNameProvided** [comment]
On the one hand, exposing class-transform internals to the host class is a no-go. It also restrict the usefulness of the class as when it's called outside of a class transformation cycle what value do we give back to it, what happens if we just transformed 100k elements, do we store all the groups for every transformation?

On the other hand, I am fine with exposing it to the `@Transform` decorator as transform functions are always called in the context of an actual transformation cycle.
